project/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -101,6 +116,87 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     return elections;
   }
 
+  /**
+   * Takes an {@code Entity} object which represents an election, and returns true if
+   * this election is deemed relevant to the given {@code address}. If the {@code listAll}
+   * parameter is true, then the election is always deemed relevant.
+   */
+  private boolean isRelevantElection(Entity election, String address, boolean listAllElections) {
+    if (listAllElections) {
+      return true;
+    } else {
+      // Query the Civic Information API for whether {@code election} is relevant to
+      // {@code address}.
+      String queryUrl =
+        String.format("%s&address=%s&electionId=%s", VOTER_INFO_QUERY_URL,
+                      address.replace(" ", "%20"), (String) election.getProperty("queryId"));
```

Comment:
I thought you were going to do a query on the address that would return a list of relevant election IDs, and then use this to filter down the elections?

Reply:
That functionality isn't provided in reality. Although electionId is technically an optional parameter and the query is supposed to return all election names/IDs relevant to an address, the query fails when electionId is not specified.

Reply:
Ah that's weird/unfortunate

Reply:
So I tried some random addresses with random elections, and it looks like it'll be successful regardless of whether the address is even in the same state as the election?

Reply:
I also tried addresses in the relevant state but could never get things like polling information to show up.  I wonder if it's because of covid or something else is broken?

Reply:
Oh! It's because I manually set the election query ID of the "New York's 14th Congressional District Election" to a magic election query ID (2000). This ID actually corresponds to a test election from the Civic Information API, which applies to the entire country. This is why that election will be deemed relevant to all addresses. I did this because I wanted something to show up in the directory page.

Reply:
Other elections are not appearing because certain election information has not been populated, eg. incumbency, candidate positions. In Gabi's earlier pull request, she added a NullPointerException checking step in the DataServlet that catches non-populated data. Those other elections that fit this criteria would be discarded, and therefore would not appear on the directory page.
"New York's 14th Congressional District Election"  can appear because it is a manually hard-coded election that didn't come from the Civic Information API; it has all the data populated. That's why it was not discarded by the DataServlet.

Reply:
Oh I was actually testing on the Civic Info site directly: https://developers.google.com/civic-information/docs/v2/elections/voterInfoQuery#try-it

Reply:
It might be because currently few elections are active right now

Reply:
Yeah I tried Georgia and Michigan primaries, but no luck with random Georgia/Michigan addresses for polling locations.  But I still got a "successful" response for these with a California address, so I'm not sure your current method of validation will actually weed out any elections.

Reply:
I see what you're saying now. I just did some tests.
I'm currently weeding elections out by checking whether the response contains a `contests` field, if it is successful. But it does seem that even relevant elections might not have the `contests` field (which contains position information). Only election 2000 has that. I might need to find a different way to do this

Reply:
I did some additional tests. It appears that for a few elections, when the address and election ID match, the "sources" field will have "Voting Information Project" as the name and will set to "official". The response will also have a "physical address". When wrong addresses are checked with these elections, these fields are different. However, there are exceptions. For instance, the WA election doesn't exhibit this behavior when the address matches.
But currently, this is the best way I can think of now to weed out elections (as opposed to checking for the "contests" field, which is present only in the perfect election 2000, and then weeding out all elections)

Reply:
As an example, note how the responses are slightly different (though sadly they both don't contain the "contests" field)
WI + Election 5009 (match)
https://developers.google.com/civic-information/docs/v2/elections/voterInfoQuery?apix=true&apix_params=%7B%22address%22%3A%22WI%22%2C%22electionId%22%3A5009%7D#try-it

WA + election 5009 (doesn't match)
https://developers.google.com/civic-information/docs/v2/elections/voterInfoQuery?apix=true&apix_params=%7B%22address%22%3A%22WA%22%2C%22electionId%22%3A5009%7D#try-it